java/arcs/core/allocator/CollectionHandlePartitionMap.kt

```diff
@@ -0,0 +1,131 @@
+package arcs.core.allocator
+
+import arcs.core.common.ArcId
+import arcs.core.data.FieldType
+import arcs.core.data.HandleMode
+import arcs.core.data.Plan
+import arcs.core.data.RawEntity
+import arcs.core.data.Schema
+import arcs.core.data.SchemaFields
+import arcs.core.data.SchemaName
+import arcs.core.entity.EntityBase
+import arcs.core.entity.EntityBaseSpec
+import arcs.core.entity.HandleContainerType
+import arcs.core.entity.HandleSpec
+import arcs.core.entity.ReadWriteCollectionHandle
+import arcs.core.entity.awaitReady
+import arcs.core.host.EntityHandleManager
+import arcs.core.storage.keys.RamDiskStorageKey
+import arcs.core.storage.referencemode.ReferenceModeStorageKey
+import arcs.core.util.TaggedLog
+import kotlinx.coroutines.joinAll
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.withContext
+
+/**
+ * An implementation of [Allocator.PartitionSerialization] that stores partition information in an Arcs
+ * collection handle, created by the [EntityHandleManager] provided at construction. The handle
+ * will be created the first time any of the publicly exposed methods is called.
+ */
+class CollectionHandlePartitionMap(
+    private val handleManager: EntityHandleManager
+) : Allocator.PartitionSerialization {
+
+    private val log = TaggedLog { "CollectionHandlePartitionMap" }
+
+    private val collectionMutex = Mutex()
+    private lateinit var _collection: ReadWriteCollectionHandle<EntityBase>
+
+    private suspend fun collection() = collectionMutex.withLock {
+        if(!::_collection.isInitialized) {
```

Comment:
Nice, I've never seen this trick before to check lateinit and not throw an exception, since 100% of my usage of lateinit has been in test setup.